profit_analysis/prices.py

```diff
@@ -0,0 +1,191 @@
+            )
+
+    async def get_price_at_block(self, block_number: Union[int, float]):
+        print(f"STARTING PRICER block_number={block_number}")
+        trials = 0
+        n_trials = 3
+        while trials < n_trials:
+            trials += 1
+            try:
+                if self._token_target_address == self._token_base_address:
+                    return 1.0
+                reserves = self._pair.functions.getReserves().call(
+                    block_identifier=int(block_number)
+                )
+                if self._is_target_token0_or_token1 == 0:
+                    token_target_reserves = reserves[0]
+                    token_base_reserves = reserves[1]
+                else:
+                    token_target_reserves = reserves[1]
+                    token_base_reserves = reserves[0]
+
+                print(
+                    f"type self._token_base_decimals={type(self._token_base_decimals)}"
+                )
+                print(f"type token_base_reserves={type(token_base_reserves)}")
+                print(f"self._token_base_decimals={self._token_base_decimals}")
+                print(f"token_base_reserves={token_base_reserves}")
+                price = (
+                    (float(token_base_reserves) / float(token_target_reserves))
+                    * self._token_target_decimals
+                    / self._token_base_decimals
+                )
+
+                return float(price)
+            except Exception as e:
+                print(
+                    f"Error ({trials}/{n_trials}), retrying get_price_at_block  - {e}"
+                )
+                sleep(0.05)
+        W3.rotate_rpc_url()
+        await self.create(self._token_target_address)
+        price = await self.get_price_at_block(block_number)
+        self.block_to_price[block_number] = price
+
+
+async def safe_get_price(pricer, block, max_concurrency):
+    async with max_concurrency:
+        print(block)
+        return await pricer.get_price_at_block(block)
+
+
+async def get_uniswap_historical_prices(
+    block_number_min, block_number_max, token_address, chain=POLYGON_CHAIN
+):
+    # TODO: get only the blocks, not every nthing
+    # print(f"pricer for token={token} from={block_number_min}, to ={block_number_max}")
+    # token_cg_ids = get_address_to_coingecko_ids_mapping("ethereum", False)
+    # token_addresses = token_cg_ids.loc[token_cg_ids[CG_ID_KEY] == cg_id, TOKEN_KEY]
+    # token_address = token_addresses.values[0]
+    if token_address != "NAN":
+        pricer = UniswapPricer(W3, chain)
+        # we use USDC as a base token
+        await pricer.create(token_address)
+        blocks = [
+            block_number_min + i
+            for i in range(int(block_number_max + 1 - block_number_min))
+        ]
+        print(blocks)
+        tasks = []
+        max_c = 1
+        max_concurrency = asyncio.Semaphore(max_c)
+        for block in range(block_number_min, block_number_max, max_c):
```

Comment:
there must be something wrong around here. Getting
`maximum recursion depth exceeded while calling a Python object` error